src/metrics.cpp

```diff
@@ -103,6 +104,30 @@ double GetLocalSolPS()
     return miningTimer.rate(solutionTargetChecks);
 }
 
+int EstimateNetHeightInner(int height, int64_t tipmediantime,
+                           int heightLastCheckpoint, int64_t timeLastCheckpoint,
+                           int64_t targetSpacing)
+{
+    // We average the target spacing with the observed spacing to the last
+    // checkpoint, and use that to estimate the current network height.
+    int medianHeight = height - CBlockIndex::nMedianTimeSpan / 2;
+    double checkpointSpacing = (double (tipmediantime - timeLastCheckpoint)) / (medianHeight - heightLastCheckpoint);
+    double averageSpacing = (targetSpacing + checkpointSpacing) / 2;
+    int netheight = medianHeight + ((GetTime() - tipmediantime) / averageSpacing);
```

Comment:
Under what conditions could `averageSpacing == 0`? Three cases to consider:

- `targetSpacing == checkpointSpacing == 0`
- One is negative (which seem unexpected).
- The other is negative.